Add initial authentication platform enhancement

scotwells opened the pull request:
### Summary

This PR creates a new enhancement document for #19 that proposes replacing the authentication system built into Datum OS with a new authentication system that leverages open source software. This new authorization system will provide us the foundation we need to support enterprise authentication options and authenticate machine users. 

A new document was also introduced to provide a high-level overview of the Identity and Access Management system being added to Datum Cloud. A document will be added in the future that covers the architecture of the policy engine component already built into the Datum Cloud IAM system.

### Goals

- Define summary, motivation, and goals of replacing the current authentication system
- Define high level system architectural goals of the authentication system
- Define authorization flows that we want to support for users and machines

### Non-Goals

- Define implementation guidance for any system components or evaluate any open source providers
- Define system architecture or components required for Workload Identity Federation or SCIM support.
- Define details on how the authorization's policy engine architecture or implementation

---

Relates to #19 
Closes #22

mustela commented on enhancements/business-os/identity-and-access-management/README.md:
````diff
@@ -0,0 +1,68 @@
+  APIServer-->|Authenticates users in|AuthProvider
+```
+
+### Authentication Platform
````
It’s not immediately clear from the diagram (at least to me) which components belong to each specific “platform.” I assume the Authentication Provider and Identity Provider are part of the Authentication Platform, while the IAM Policy Engine is part of the Authorization Platform. But I’m unsure about the role of the IAM API Server—is it just a gateway? Does it sit in front of both the Authentication and Authorization Platforms?

Also, the term “Platform” feels a bit ambiguous (again, maybe just for me). What exactly are you envisioning with that term?

scotwells replied:
This diagram is really just illustrating the IAM platform and the components that make it up, it doesn't break it down into the authentication and authorization platform. 

Maybe we remove the "platform" from Authentication and Authorization to make it less confusing since that's really capabilities the IAM platform provides? We could also consider referring to it as the IAM system since it's a system of components that work together to provide a set of functionality. 

The IAM API Server is what provides the IAM functionality that aligns with our platform. So it should enable things like:
- User management in an organization
- Service account management in a project
- IAM policy management for all resources
- IAM role / permission management for services
- IAM access check functionality

I could update this document with a "components" section that breaks down the functionality we'd expect each component to have if you think that would be helpful.

mustela replied:
> Maybe we remove the "platform" from Authentication and Authorization to make it less confusing since that's really capabilities the IAM platform provides? 

Agree 100% 🙏🏼 

> I could update this document with a "components" section that breaks down the functionality we'd expect each component to have if you think that would be helpful.

That would be awesome! It doesn't have to be very detailed, but what you wrote about the IAM API Server is a great example.

scotwells replied:
@mustela just pushed some updates to remove "authentication platform" from the authentication enhancement / IAM overview. Also pushed an update to include a component breakdown that covers the APIServer.

mustela commented on enhancements/business-os/identity-and-access-management/authentication-platform/README.md:
```diff
@@ -0,0 +1,804 @@
+    Service
+  end
+
+  Machines
```
"Machines" are basically microservices, third-party APIs. Basically non-human entity, right?

scotwells replied:
That's correct

mustela reviewed (dismissed):
LGTM!

scotwells commented:
@mustela FYI I updated the PR to use the C4 diagram that I already had in the Datum OS repo for the design of the IAM system. I'll update this diagram later to include the authentication and identity provider used to authenticate users/machines.

I also added some details around the concepts that we expose to users through our IAM system and examples of what those resources look like.

scotwells commented on enhancements/business-os/identity-and-access-management/README.md:
```diff
@@ -0,0 +1,68 @@
+
+Roles are also inherited when IAM policies are bound to parent resources like
+Organizations and Projects. Read through the [service provider
+platform enhancement](../service-provider-platform/) for more information.
```
@JoseSzycho good catch! Just pushed up a fix for that.
